pkg/generator/generator.go

```diff
@@ -0,0 +1,140 @@
+package generator
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/template"
+)
+
+type Element interface {
+	Name() string
+	Template() string
+	Create(*template.Template) *template.Template
+	Data() interface{}
+}
+
+type Elements []Element
+
+type Section struct {
```

Comment:
Hierarchy and indenting: we may need more than 2 levels of hierarchy in future.
Suggestion: Make Section implement the Element interface. Then we can have sections within sections with correct indenting. Here's a sketch:
```
package main

import (
	"bufio"
	"fmt"
	"strings"
)

type Element interface {
	Template() string
}

func indentTemplates(indent int, es []Element) string {
	pad := strings.Repeat(" ", indent)
	b := &strings.Builder{}
	for _, e := range es {
		scanner := bufio.NewScanner(strings.NewReader(e.Template()))
		for scanner.Scan() {
			fmt.Fprintf(b, "%v%v\n", pad, scanner.Text())
		}
	}
	return b.String()
}

type Section []Element

func (s Section) Template() string { return indentTemplates(0, []Element(s)) }

type Indented []Element

func (s Indented) Template() string { return indentTemplates(2, []Element(s)) }

// Dummy element, just a string
type Text struct{ Text string }

func (t Text) Template() string { return t.Text }

func main() {
	conf := Section{
		Text{"hello"},
		Indented{
			Text{"foo"},
			Indented{
				Text{"inside"},
			},
			Text{"bar"},
		},
		Text{"bye"},
	}
	fmt.Println(conf.Template())
}

```

Reply:
`Section` is a convenience type, all the heavy lifting is done by `Element` and `[]Element` 
`Section` is used only in the first level, to show a breakdown of generated configuration + add a comment along with the declared section to document the meaning of the section.
e.g.
```
{
	PrometheusMetrics(clfspec, op),
	"Section to add measurement, and dispatch to Concat or Ingress pipelines",
},
{
	Concat(clfspec, op),
	`Concat pipeline 
	section`,
},
```
Need your and @jcantrill help to fill up the comments for each `Section`

Reply:
for
```
type Text struct{ Text string }
func (t Text) Template() string { return t.Text }
```

we already have a type `ConfLiteral` which is used to combine a template defined as a string
https://github.com/openshift/cluster-logging-operator/pull/1118/files#diff-f589689c1969e276096b90ed4bf632a068e506f1509ffafe8d9dc3273239a775
usage: https://github.com/openshift/cluster-logging-operator/pull/1118/files#diff-5d532129029d7d81e3ad40ecb9e6b36b09c30771e0727ba74a04009197141f0cR32

Reply:
About multiple levels, we already support that with Element